### PR TITLE
[WIP] fix libunwind

### PR DESCRIPTION
**Pull request by aleksmelnikov**
fix compilation panic=abort for target i686-pc-windows-gnu on linux with w64-mingw32-gcc linker

### CODE REVIEW

File: src/libunwind/libunwind.rs [hunk 1]
```diff
@@ -212,32 +211,55 @@ if #[cfg(all(any(target_os = "ios", target_os = "netbsd", not(target_arch = "arm
...
+    // 32-bit iOS
+    if #[cfg(all(target_os = "ios", target_arch = "arm"))] {
+        pub use sjlj::*;
+    }
+    // FIXME: for compatibility with i686-w64-mingw32-gcc on linux
+    else if #[cfg(all(target_vendor = "pc",
+                      target_os = "windows",
+                      target_arch = "x86",
+                      target_env = "gnu"))] {
+        pub use sjlj::*;
```
Comment (mati865) [inline src/libunwind/libunwind.rs]: Won't this would change behaviour of the target resulting in breaking change?

**Comment by alexcrichton** [comment]
Thanks for the PR! This PR is unfortunately pretty light on the details. Neither the PR title nor description really gives any insight into what the compilation error looks like today, what this change is, and why the change fixes it. Could you expand the descriptions and/or give more detail about the issues you're encountering?

**Comment by superriva** [comment]
Any breaking changes is not what I want it all.
This fix:
1. to add modules for dwarf2 and sjlj exception type for platforms
2. move i686-pc-windows-gnu from dwarf2 to sjlj for cross-compiling because mingw package on linux supports sjlj for i686

**Comment by superriva** [comment]
https://github.com/rust-lang/rust/issues/32859

**Comment by retep998** [comment]
Rust currently ships `i686-pc-windows-gnu` with a MinGW toolchain that uses dwarf exception handling, not sjlj. Wouldn't your change conflict with that?

File: src/libunwind/libunwind.rs (hunk 1 above)
Comment (superriva) [reply src/libunwind/libunwind.rs]: @retep998 , yes you are right. The patch conflicts with that and must be changed.

**Comment by superriva** [comment]
[the first atempt to fixing](https://github.com/rust-lang/rust/pull/49633).

>After some more testing, I've discovered this approach will not work with Rust as-shipped, as the pre-compiled stdlib binaries shipped to rustup are compiled on their native platforms. With the way Rust is currently distributed, I don't think fixing this will be possible at the moment (until Xargo functionality is merged into Cargo, most likely).

**Comment by alexcrichton** [comment]
Ok thanks for expanding the description! As you've discovered I don't think this is fixable without rustc changes itself. Even with xargo the compiler needs to configure LLVM and may need to change its codegen for sjlj exceptions.

Do you want to leave this open in the meantime or shall we close this?

**Comment by superriva** [comment]
@alexcrichton , I do not throw this PR. I look through some more possible solutions. I would like the solution to be as minimal and simple as possible. 
I also contacted the `debian` `mingw` package mantainer. And it seemed to me he looks at the `sjlj` more positively than on the `dwarf`. Why `rust` choose `dwarf2` exception on `i686`, but not `sjlj` ?

**Comment by alexcrichton** [comment]
This probably isn't the best place to discuss why architectural design decisions are the way they are in Rust, I'd recommend starting a thread on internals.rust-lang.org for that. I don't think this is an actionable change without more changes to the compiler itself, so I'm inclined to close this while those are settled.

**Comment by superriva** [comment]
@alexcrichton 

> I don't think this is an actionable change without more changes to the compiler itself

Yes, more changes are needed in compiler code and scripts.
Now I tested much more code changes at my hosting and I can compile new target to stage1, stage2 gives errors. Further work in this direction is not needed?

**Comment by superriva** [comment]
Probably adding new targets for compile features is not good path.

**Pull request closed without merging**